lib/plugins/sensorage.js

```diff
@@ -0,0 +1,173 @@
+      , debug: {
+        sensorInfo: sensorInfo
+      }
+    });
+  }
+};
+
+function minButValid(record) {
+  if (!record['Sensor Start'].age) {
+    return 'Sensor Change';
+  }
+  if (!record['Sensor Change'].age) {
+    return 'Sensor Start';
+  }
+  if (record['Sensor Start'].age < record['Sensor Change'].age) {
+    return 'Sensor Start';
+  }
+  if (record['Sensor Start'].age > record['Sensor Change'].age) {
+    return 'Sensor Change';
+  }
+  alert('Some bug');
+}
+
+sage.findLatestTimeChange = function findLatestTimeChange(sbx) {
+
+  var returnValue = {
+    'Sensor Start': {'message':'', 'found': false, 'age': 0, 'treatmentDate': null, 'checkForAlert': false}
+    , 'Sensor Change': {'message':'', 'found': false, 'age': 0, 'treatmentDate': null, 'checkForAlert': false}
+  };
+
+  var prevDate = {
+    'Sensor Start': 0
+    ,'Sensor Change': 0
+  };
+
+  _.each(sbx.data.sensorTreatments, function eachTreatment (treatment) {
+    ['Sensor Start', 'Sensor Change'].forEach( function eachEvent(event) {
+      var treatmentDate = treatment.mills;
+      if (treatment.eventType === event && treatmentDate > prevDate[event] && treatmentDate <= sbx.time) {
+
+        prevDate[event] = treatmentDate;
+        returnValue[event].treatmentDate = treatmentDate;
+
+        //allow for 10 minute period after a full hour during which we'll alert the user
+        var a = moment(sbx.time);
+        var b = moment(returnValue[event].treatmentDate);
+        var days = a.diff(b,'days');
+        var hours = a.diff(b,'hours') - days * 24;
+        var age = a.diff(b,'hours');
+        var minFractions = a.diff(b,'minutes') - age * 60;
+        
+        returnValue[event].checkForAlert = minFractions <= 10;
+
+        if (!returnValue[event].found) {
+          returnValue[event].found = true;
+          returnValue[event].age = age;
+          returnValue[event].days = days;
+          returnValue[event].hours = hours;
+        } else {
+          if (age >= 0 && age < returnValue[event].age) {
+            returnValue[event].age = age;
+            returnValue[event].days = days;
+            returnValue[event].hours = hours;
+            if (treatment.notes) {
+              returnValue[event].message = treatment.notes;
+            } else {
+              returnValue[event].message = '';
+            }
+          }
+        }
+      }
+    });
+  });
+  
+  // clear Sensor Start if both found and Sensort Change is newer
+  if (returnValue['Sensor Change'].found && returnValue['Sensor Start'].found && returnValue['Sensor Start'].treatmentDate < returnValue['Sensor Change'].treatmentDate) {
+    returnValue['Sensor Start'] = {'message':'', 'found': false, 'age': 0, 'treatmentDate': null, 'checkForAlert': false};
+  }
+  returnValue.min = minButValid(returnValue);
+
+  return returnValue;
+};
+
+sage.updateVisualisation = function updateVisualisation (sbx) {
+
+  var sensorInfo = sage.findLatestTimeChange(sbx);
+  var prefs = sage.getPrefs(sbx);
```

Comment:
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: ['prefs' is defined but never used.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1269677074/issues/source?bid=2923525&fileBranchId=2942205#l140)